[WIP] higher level structured clone API

alajara opened the pull request:
I ran build -d and it  builds and compiles but when I run the test t./mach test structured clone tests one of the tests crashes.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ X] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [x] These changes fix #16031 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16321)
<!-- Reviewable:end -->

Commit:
[WIP]

I ran build -d and it  builds and compiles but when I run the test t./mach test structured clone tests one of the tests crashes.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @cbrewster (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/bindings/structuredclone.rs
 * @kichjang: components/script/dom/bindings/structuredclone.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,49 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+}
+
+impl StructuredCloneReader {
+    unsafe fn read_bytes(&self) {
```
This method needs to return the bytes the bytes that are read.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,49 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+
+impl StructuredCloneReader {
+    unsafe fn read_bytes(&self) {
+        assert!(JS_ReadBytes(self.r, vec![0u8; self.read_usize()].as_mut_ptr() as *mut raw::c_void, self.read_usize()));
```
This reads two separate usize values, rather than reading one and then using that value as the vector length as well as number of bytes to read. Please use a local variable instead.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,49 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+    unsafe fn read_usize(&self) -> usize {
+        read_length(self.r)
+    }
+    unsafe fn read_slice(&self) -> Vec<u8> {
```
This method should not be necessary.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,49 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+        assert!(JS_ReadBytes(self.r, vec![0u8; self.read_usize()].as_mut_ptr() as *mut raw::c_void, self.read_usize()));
+    }
+    unsafe fn read_str(&self) -> String {
+        String::from_utf8_unchecked(vec![0u8; self.read_usize()];
```
This doesn't read any bytes for the string.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,49 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+
+impl StructuturedCloneWriter {
+    unsafe fn write_slice(&self, v: &[u8]) {
+        assert!(JS_WriteBytes(self.w, v.as_ptr() as *const raw::c_void, v.len()));
```
We should be writing the length before writing the bytes, so the calling code does not need to write the length manually.

alajara replied:
so should I call write_length function in CloneWriter Struct and pass that to a variable then use that for length?

jdm reviewed (changes requested):
This is the right idea, but the point of the `write` APIs is the use higher level concepts that encapsulate things like "write a length followed by that many bytes". Similarly, when reading there should be no need to use the "read the length" followed by the "read the bytes" operations - things like `read_bytes` and `read_str` will encapsulate those, so the calling code can just say "read a string".

jdm commented:
@alajara Are you planning to keep working on this?

alajara commented:
Yes sorry I was away for awhile, I see that the you commented on what I need to correct, I will start correcting it.

Commit:
Update structuredclone.rs

I create the local variable in read functions and separated the Read bytes and assert.

Commit:
Update structuredclone.rs

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+}
+
+impl StructuredCloneReader {
+    unsafe fn read_bytes(&self) -> usize {
```
`read_bytes` is a fundamental operation that should:
* read a length value N
* create a vector of size N
* read N bytes into that vector
* return the vector

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+        // return string from vec!
+        let type_str_length = read_length(self.r);
+        let mut type_str_buffer = vec![0u8; type_str_length];
+        return String::from_utf8_unchecked(type_str_buffer);
```
This method should invoke `read_bytes` and convert the vector that it returns into a string.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+        let mut type_str_buffer = vec![0u8; type_str_length];
+        return String::from_utf8_unchecked(type_str_buffer);
+    }
+    unsafe fn read_slice(&self) -> Vec<u8> {
```
This method should not be necessary, since it doesn't not provide a benefit over read_bytes.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+        let vect = vec![0u8; blob_buffer];
+        return vect;
+    }
+    unsafe fn read_assert(&self, v: &mut [u8], l: usize) {
```
It is not clear what this method is supposed to do that is not handled by the previous methods.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+    structured_reader.read_assert(&mut blob_buffer, blob_length);
+    let mut type_str_buffer = structured_reader.read_slice();
+    structured_reader.read_assert(&mut type_str_buffer, type_str_length);
+    let type_str = structured_reader.read_str();
```
Given the changes requested previously, I would expect this code to look like:
```rust
let blob_buffer = structured_reader.read_bytes();
let type_str = structured_reader.read_str();
```
That's all.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+impl StructuredCloneWriter {
+    unsafe fn write_slice(&self, v: &[u8]) {
+        // writers bytesl takes in a vec! and its length
+        let blob_length = v.len();
```
This method should write the length of the slice as well.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -85,15 +131,16 @@ unsafe fn read_blob(cx: *mut JSContext,
+    structured_writer.write_usize(&blob_length);
+    structured_writer.write_usize(&type_string_length);
+    structured_writer.write_slice(&blob_vec);
+    structured_writer.write_slice(&type_string_bytes);
```
This code should not have to write the lengths explicitly; that should be the responsibility of `write_slice`.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -85,15 +131,16 @@ unsafe fn read_blob(cx: *mut JSContext,
+    let structured_writer = StructuredCloneWriter{w: w};
     let blob_vec = try!(blob.get_bytes());
     let blob_length = blob_vec.len();
     let type_string_bytes = blob.get_type_string().as_bytes().to_vec();
```
The `to_vec()` is not necessary any more.

Commit:
Update structuredclone.rs

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -72,56 +72,36 @@ struct StructuredCloneWriter {
 impl StructuredCloneWriter {
-    unsafe fn write_slice(&self, v: &[u8]) {
-        // writers bytesl takes in a vec! and its length
+    unsafe fn write_slice(&self, s: &usize, v: &[u8]) {
```
We should not need to take an extra argument for the size. We should use `v.len()`.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -72,56 +72,36 @@ struct StructuredCloneWriter {
-        return type_length;
+        let mut v = vec![0u8; type_length];
+        let blob_length = v.len();
+        assert!(JS_ReadBytes(self.r, v.as_mut_ptr() as *mut raw::c_void, blob_length));
```
I was confused when I first read this. I would be less confused if we kept using `type_length` everywhere instead of another variable.

jdm reviewed (changes requested):
Almost there!

Commit:
Update structuredclone.rs

alajara commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -85,15 +110,12 @@ unsafe fn read_blob(cx: *mut JSContext,
-    assert!(JS_WriteBytes(w, blob_vec.as_ptr() as *const raw::c_void, blob_length));
-    assert!(JS_WriteBytes(w, type_string_bytes.as_ptr() as *const raw::c_void, type_string_length));
+    structured_writer.write_slice(&blob_vec);
+    structured_writer.write_slice(&type_string_bytes);
```
Should I remove blob.get_type_string().as_bytes().to_vec();?,  I am not sure how to get the type_string without calling it. thanks for the help.

jdm replied:
No. I just meant that the `to_vec()` call is unnecessary; the rest of the expression is still important.

Commit:
Merge branch 'master' into patch-3

Commit:
Merge branch 'master' into patch-3

Commit:
remove to_vec() call

alajara commented:
Thanks for the guidance, I really appreciate it.

Commit:
Update structuredclone.rs

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,41 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+        let type_length = v.len();
+        write_length(self.w, type_length);
+        assert!(JS_WriteBytes(self.w, v.as_ptr() as *const raw::c_void, type_length));
+    }
```
Let's add a corresponding `write_str` which accepts a `&str` argument and calls `write_slice(s.as_byte())`.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -85,15 +110,11 @@ unsafe fn read_blob(cx: *mut JSContext,
-    assert!(JS_WriteBytes(w, blob_vec.as_ptr() as *const raw::c_void, blob_length));
-    assert!(JS_WriteBytes(w, type_string_bytes.as_ptr() as *const raw::c_void, type_string_length));
+    structured_writer.write_slice(&blob_vec);
+    structured_writer.write_slice(blob.type_string().as_bytes(););
```
This doesn't compile as written. We can also make use of the `write_str` method that I propose above using `structured_writer.write_str(&blob.type_string());`.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,41 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+
+impl StructuredCloneReader {
+    unsafe fn read_bytes(&self) -> Vec<u8> {
+        let mut type_length = vec![0u8; read_length(self.r)];
```
Let's rename this `bytes` instead of `type_length`, since it's a buffer and not a length.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -67,16 +67,41 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
+        return type_length;
+    }
+    unsafe fn read_str(&self) -> String {
+        let type_str_buffer = self.read_bytes();
```
Let's rename this `str_buffer`.

Commit:
Update structuredclone.rs

jdm commented:
This looks great! Could you [squash](https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/) the commits into one and give them a meaningful commit message like `Create a high level API for performing structured clone operations.`?

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit e24e0f24e375e6383a876ea70b8a1634d94a8a0e with merge 6c98d010015525f5b3a6af7f67c65e71167acda4...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/7054)

cbrewster commented:
Fails tidy:
```
Checking files for tidiness...

./components/script/dom/bindings/structuredclone.rs:105: missing space before {

./components/script/dom/bindings/structuredclone.rs:105: missing space before }

./components/script/dom/bindings/structuredclone.rs:105: missing space after {

./components/script/dom/bindings/structuredclone.rs:116: missing space before {

./components/script/dom/bindings/structuredclone.rs:116: missing space before }

./components/script/dom/bindings/structuredclone.rs:116: missing space after {
```

alajara commented:
still working on this, I am still a novice when it comes to using git on the command line, but I am figuring it out, I am trying to get the this branch, in a my repository to squash the other commits.

jdm commented:
Given that #16846 got messed up with a bunch of other unrelated Servo changes, I went ahead and rebased this into a single change. I cherry-picked each commit (excluding the merges), fixed up a couple resulting merge conflicts, and then did an interactive rebase. These are the commands I ran :
```
$ git checkout master
$ git checkout -b structuredclone
$ git cherry-pick 3d80254
$ git cherry-pick 8d9743c
$ git cherry-pick 2a02f74
$ git cherry-pick 1baab4d
$ git cherry-pick 8446166
... addressed merge conflict ...
$ git add components/script/dom/bindings/structuredclone.rs
$ git cherry-pick --continue
$ git cherry-pick acf8ae9
... addressed merge conflict ...
$ git add components/script/dom/bindings/structuredclone.rs
$ git cherry-pick --continue
$ git cherry-pick 28c3a11
$ git cherry-pick e24e0f2
$ git rebase -i 52640cd46ea83e2abb139eb664d9d01a65b9f839^
```

jdm commented:
Rebased in #16884.
